src/app/error.rs

```diff
@@ -0,0 +1,75 @@
+use thiserror::Error;
+
+use crate::database;
+
+#[derive(Debug, Error)]
+pub enum InsertIdentityV2Error {
+    #[error("provided identity commitment is invalid")]
+    InvalidCommitment,
+    #[error("provided identity commitment is not in reduced form")]
+    UnreducedCommitment,
+    #[error("provided identity commitment is already included")]
+    DuplicateCommitment,
+    #[error("provided identity commitment was deleted")]
+    DeletedCommitment,
+    #[error(transparent)]
+    Database(#[from] database::Error),
+    #[error(transparent)]
+    Sqlx(#[from] sqlx::Error),
+}
+
+#[derive(Debug, Error)]
+pub enum DeleteIdentityV2Error {
+    #[error("provided identity commitment is invalid")]
+    InvalidCommitment,
+    #[error("provided identity commitment is not in reduced form")]
+    UnreducedCommitment,
+    #[error("provided identity commitment is not yet added to the tree")]
+    UnprocessedCommitment,
+    #[error("provided identity commitment was not found")]
+    CommitmentNotFound,
+    #[error("provided identity commitment was deleted")]
+    DeletedCommitment,
+    #[error("provided identity commitment already scheduled for deletion")]
+    DuplicateCommitmentDeletion,
+    #[error(transparent)]
+    Database(#[from] database::Error),
+    #[error(transparent)]
+    Sqlx(#[from] sqlx::Error),
+}
+
+#[derive(Debug, Error)]
+pub enum InclusionProofV2Error {
+    #[error("provided identity commitment is invalid")]
+    InvalidCommitment,
+    #[error("provided identity commitment is not yet added to the tree")]
+    UnprocessedCommitment,
+    #[error("provided identity commitment is not in reduced form")]
+    UnreducedCommitment,
+    #[error("provided identity commitment was not found")]
+    CommitmentNotFound,
+    #[error("provided identity commitment was deleted")]
+    DeletedCommitment,
+    #[error(transparent)]
+    Database(#[from] database::Error),
+    #[error(transparent)]
+    Sqlx(#[from] sqlx::Error),
+    #[error(transparent)]
+    AnyhowError(#[from] anyhow::Error),
+}
+
+#[derive(Debug, Error)]
+pub enum VerifySemaphoreProofV2Error {
+    #[error("provided root is invalid")]
+    InvalidRoot,
+    #[error("provided root is too old")]
+    RootTooOld,
+    #[error("cannot decompress provided proof")]
+    DecompressingProofError,
+    #[error("prover error")]
+    ProverError,
+    #[error(transparent)]
+    Database(#[from] database::Error),
+    #[error(transparent)]
+    AnyhowError(#[from] anyhow::Error),
```

Comment:
A friend enlightened me recently to a cool pattern, where instead of doing this, we go for:

```suggestion
    RootValidityError(anyhow::Error),
```

which makes it a little more difficult to use but it makes the choice of the error explicit. So at the error propagation site we have to do something like

```rust
let is_root_valid = self.validate_root_age(root_age)
    .map_err(VerifySemaphoreProofV2Error::RootValidityError)?;
```